### DISCUSSION TITLE
How to calculate PCoA with Bray-Curtis

### DISCUSSION THREAD

**Discussion by jonng2024** [scikit-bio/scikit-bio#2081] [Q&A] [answered]
Hi, I'm new to microbiome analysis and using sckit-bio. I would like to perform PCoA using the Bray-Curtis distance. May I know how to do this please? Confused even after looking at the documentation.

```
from skbio.stats.ordination import pcoa
from scipy.spatial.distance import pdist, squareform

# Calculating Bray-Curtis distance
# sc_data has my bacterial taxa data
bc_dist = pdist(sc_data, 'braycurtis')
distance_matrix = squareform(bc_dist)

#PCoA
pcoa_result = pcoa(distance_matrix)
```
Would the above be correct?

Do I need to use this code `skbio.diversity.beta_diversity` ? If so, may I kow how to use it please? Thank you

**Comment by wasade** [comment]
Hi @jonng2024, thank you for reaching out. What I think is missing here is expressing the `distance_matrix` as a `skbio.DistanceMatrix` type. It should be possible to pass in the current array, as well as the sample IDs and then use that distance matrix with `pcoa`. That said, using the `beta_diversity` method would do that automatically and for this metric, if I recall correctly, is just using scipy and `pdist`

**Reply by jonng2024** [reply]
Ok great thanks! =)

**Comment by qiyunzhu** [comment]
@jonng2024 There is actually an example in the [scikit.bio](https://scikit.bio) homepage. Basically, you will need:

```python
from skbio.diversity import beta_diversity
from skbio.stats.ordination import pcoa

distance_matrix = beta_diversity('braycurtis', sc_data, ids=sample_ids)
pcoa_result = pcoa(distance_matrix)
```

**Reply by jonng2024** [reply]
Ok great thanks! =). Found the example at https://scikit.bio/docs/latest/diversity.html#beta-diversity

**Reply by jonng2024** [reply]
Hi its me again, I tried your code, but it doesnt work on my microbiome dataset, which has n samples with m features
This dataset has undergone multiplicative replacement and clr transformation (using scikit-bio's functions). Next it was transformed and normalised using QuantileTransformer and StandardScaler from sckit-learn. Thus, it has some negative values.

The code I used is:
```
from skbio.diversity import beta_diversity
from skbio.stats.ordination import pcoa

# taxa_df is the dataset after undergoing multiplicative replacement and clr transformation, it has my sample and feature labels
# sc_data is taxa_df after undergoing transformation and normalisation

distance_matrix = beta_diversity('braycurtis', sc_data, ids=taxa_df.index)
```
However I got:
```
ValueError: Counts cannot contain negative values.
```

I solved the issue by directly calculating the pairwise distances using scipy. May I please confirm that my code is logically correct? Would like to make sure I didn't make any mistake =)
 
```
# Calculating Bray-Curtis distance
bc_dist = pdist(sc_data, 'braycurtis')
# Making PCoA object
pcoa_result = pcoa(bc_dist)
pcoa_df = pcoa_result.samples.copy()
```

**Comment by qiyunzhu** [comment, marked as answer]
@jonng2024 I think what you suggested is at least practical. scikit-bio was designed for community data, which are supposed to be non-negative integers. But the vector distance matrices implemented in SciPy are actually more flexible to the input data. The code you wrote should work. Note that skbio's `pcoa` can take a condensed matrix (1D array) generated by `pdist` as input, although this behavior hasn't been documented.

On the other hand, it will be your call to evaluate whether it is justified to calculate Bray-Curtis dissimilarity on transformed data. This metric was originally designed for ecology, and should only take count data. I might hesitate on this. Instead, directly running PCA (not PCoA) without calculating beta diversity might be more justifiable on transformed data.

**Reply by jonng2024** [reply]
Thanks so much @qiyunzhu ! Really appreciate your hard work in answering me, it helps me understanding things a lot better =). 

**Point 1**
So yes I understand my above code would work. My sample df has 342 samples and 127 features. So logically, there should be only 127 PCs right? However` pcoa_result = pcoa(bc_dist) `produces 342 PCs. I investigated and realised the following

```
>>> pcoa_result.eigvals
PC1      80.023665
PC2      26.915290
PC3      19.406098
PC4      12.191352
PC5      11.096194
           ...    
PC338     0.000000
PC339     0.000000
PC340     0.000000
PC341     0.000000
PC342     0.000000

>>> pcoa_result.samples
          PC1       PC2       PC3       PC4       PC5       PC6       PC7       PC8       PC9      PC10  ...  PC333  PC334  PC335  PC336  PC337  PC338  PC339  PC340  PC341  PC342
0   -0.337687 -0.512459 -0.051820 -0.217247 -0.152246 -0.038011 -0.052218  0.042377 -0.137714 -0.127120  ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0
1   -0.230066 -0.486722 -0.082943  0.083018 -0.169620 -0.051632 -0.091912  0.045833 -0.130213  0.138612  ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0
2   -0.656035 -0.356773 -0.286656 -0.064383 -0.116261 -0.128773 -0.167308  0.145881 -0.073150  0.005377  ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0
3   -0.370089 -0.588169  0.057011 -0.061899 -0.259054  0.038457  0.101090 -0.080189 -0.097728 -0.084990  ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0
4   -0.723333 -0.495981 -0.035513  0.177054 -0.114167 -0.158484  0.005030  0.114859 -0.075488 -0.118053  ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0
..        ...       ...       ...       ...       ...       ...       ...       ...       ...       ...  ...    ...    ...    ...    ...    ...    ...    ...    ...    ...    ...
337  0.424751  0.409249 -0.166914 -0.046515  0.504594  0.033422 -0.172298  0.141547  0.211690  0.036492  ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0
338 -0.579490  0.241581  0.128088  0.014483 -0.218724 -0.106685 -0.017115 -0.030553 -0.247408 -0.089444  ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0
339 -0.753988  0.265215  0.263771  0.044763  0.019966 -0.119935 -0.133589  0.081762 -0.009184  0.032591  ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0
340  0.360666  0.267609 -0.356417  0.006396  0.318729 -0.118475 -0.034314  0.095888  0.202075 -0.040168  ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0
341  0.353100  0.245198 -0.122944 -0.087918  0.311273  0.187975 -0.103595  0.135187  0.066904 -0.227115  ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0

[342 rows x 342 columns]

>>> pcoa_result.proportion_explained
PC1      0.236088
PC2      0.079406
PC3      0.057252
PC4      0.035967
PC5      0.032736
           ...   
PC338    0.000000
PC339    0.000000
PC340    0.000000
PC341    0.000000
PC342    0.000000
Length: 342, dtype: float64
```
I checked the above, and anything after the 143th row/column has a value of 0. Is this behaviour expected? I am wondering if these extra 0 values would affect downstream work, such as constructing biplots, which is linked to my question #2126.

I also tested your code
```
distance_matrix = beta_diversity('braycurtis', sc_data, ids=sample_ids)
pcoa_result = pcoa(distance_matrix)
```
However, I observed the same issue,

**Point 2**
Thanks for pointing out that Bray-Curtis is meant for only positive valued data. Microbiome data is compositional, so I am required to remove its compositional constraint, and convert it into existing in euclidean space, via methods like center log ration (clr) transform. Once that happens, I think I will get negative values. I see Bray Curtis being used in the literature for clr transformed data, so I think it is possibly apporpriate.

Yes, you also have a point. An alternative way for me would be to use PCA on such transformed data, which would mean I am using PCoA using the Aitchison's distance.

On another note, I know that sckit-bio is under active development, if you feel my qns/comments are useful to others, I am happy to assist in adding them to the sckit-bio documentation

**Reply by qiyunzhu** [reply]
For your point 1: If you run PCA, the maximum dimensions is the lesser of sample number and feature number. However, PCoA is different. PCoA falls into the category of multidimensional scaling, which starts with a distance matrix instead of the original data table. For a distance matrix, it is already n x n (n is the sample number). Therefore, the dimensions returned is equal to the sample number.

For point 2, I have my thoughts but let me ping our compositionality expert @mortonjt if he wants to comment.

**Reply by jonng2024** [reply]
Ok thanks =). Will reply if there's any further updates here

**Comment by ZING-ZMJ** [comment]
Hi all, 

I‘m new to metagenomics analysis. I encountered an issue while performing the beta diversity analysis of microbiome data.  My code is:

`bc_dm = pdist(species_T.values, metric='braycurtis')` # calculate Bray-Curti's distance matrix
`dm = DistanceMatrix(bc_dm, ids=species_T.index)` #transform to 2D matrix with sample ID
`pcoa_results = pcoa(dm) ` # PCoA dimension reduction

when running this line`pcoa_results = pcoa(dm) ` it reports warning as below.
`path/py3.9env/lib/python3.9/site-packages/skbio/stats/ordination/_principal_coordinate_analysis.py:258: RuntimeWarning: The result contains negative eigenvalues that are large in magnitude, which may suggest result inaccuracy. See Notes for details. The negative-most eigenvalue is -0.24831343770764946 whereas the largest positive one is 10.030522136082013.
  warn(`

I checked the data, there are several outliers in each species, for example 
`<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/mengjing/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/mengjing/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>

</style>
</head>

<body link="#467886" vlink="#96607D">


clade_name | 0
-- | --
s__Bacteroides_fragilis | 5
s__Bacteroides_thetaiotaomicron | 3
s__Parabacteroides_distasonis | 3
s__Flavonifractor_plautii | 8
s__Collinsella_aerofaciens | 7
s__Clostridioides_difficile | 3
s__Bifidobacterium_longum | 6
s__Faecalibacterium_prausnitzii | 1



</body>

</html>
`

My question is: **Is this warning caused by these outliers? Regarding this warning on the eigenvalue, will it affect the downstream analysis?**

Thank you in advance！

**Reply by qiyunzhu** [reply]
Hello @ZING-ZMJ Good question! I'd say that you can just ignore this warning because it is too frequent on real-world data. There is a `warn_neg_eigval` parameter to control when a warning should be raised. The default value is 0.01, meaning that it will warn only when  the negative most eigenvalue exceeds 1% of the positive most eigenvalue. In your case, it is 0.24831 / 10.03052 = 2.48%, which is not too far from 1%. So don't worry and just proceed with downstream analysis. But let me also ping @mortonjt if he has any comment.
